File: easy/level_order_traversal.go
package easy

func LevelOrderTraversal(root *TreeNode) [][]int {
	result := [][]int{}
	if root == nil {
		return result
	}

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		size := len(queue)
		current := make([]int, 0, size)
		for range size {
			node := queue[0]
			queue = queue[1:]
			current = append(current, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		result = append(result, current)
	}
	return result
}

File: lib/medium/level_order_traversal.dart
import 'package:dark_coding_challenges/helpers/data_structures/tree_node.dart';

List<List<int>> levelOrderTraversal(TreeNode? root) {
  if (root == null) return [];
  List<List<int>> result = [];
  List<TreeNode?> stack = [root];
  while (stack.isNotEmpty) {
    List<int> current = [];
    int size = stack.length;
    for (int i = 0; i < size; i++) {
      TreeNode? node = stack.removeAt(0);
      int? value = node?.val;
      if (value != null) current.add(value);
      if (node?.left != null) stack.add(node?.left);
      if (node?.right != null) stack.add(node?.right);
    }
    result.add(current);
  }

  return result;
}
